File: java/com/joshmanisdabomb/lcc/block/NetherReactorBlock.java
package com.joshmanisdabomb.lcc.block;

import com.joshmanisdabomb.lcc.tileentity.NetherReactorTileEntity;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.material.MaterialColor;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.state.EnumProperty;
import net.minecraft.state.StateContainer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.ActionResultType;
import net.minecraft.util.Direction;
import net.minecraft.util.Hand;
import net.minecraft.util.IStringSerializable;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.BlockRayTraceResult;
import net.minecraft.util.text.TranslationTextComponent;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.World;

import javax.annotation.Nullable;

public class NetherReactorBlock extends Block {

    public static final EnumProperty<ReactorState> STATE = EnumProperty.create("state", ReactorState.class);

    public static final BlockState SHELL = Blocks.COBBLESTONE.getDefaultState();
    public static final BlockState BASE = Blocks.GOLD_BLOCK.getDefaultState();

    public NetherReactorBlock(Properties properties) {
        super(properties);
        this.setDefaultState(this.stateContainer.getBaseState().with(STATE, ReactorState.READY));
    }

    @Override
    protected void fillStateContainer(StateContainer.Builder<Block, BlockState> builder) {
        builder.add(STATE);
    }

    @Nullable
    @Override
    public TileEntity createTileEntity(BlockState state, IBlockReader world) {
        return new NetherReactorTileEntity();
    }

    @Override
    public boolean hasTileEntity(BlockState state) {
        return true;
    }

    @Override
    public MaterialColor getMaterialColor(BlockState state, IBlockReader world, BlockPos pos) {
        switch (state.get(STATE)) {
            case ACTIVE:
                return MaterialColor.TNT;
            case USED:
                return MaterialColor.BLACK;
            default:
                return MaterialColor.CYAN;
        }
    }

    @Override
    public ActionResultType onBlockActivated(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand handIn, BlockRayTraceResult hit) {
        if (state.get(STATE) == ReactorState.READY) {
            if (!world.isRemote) {
                if (pos.getY() < 4 || pos.getY() > 221) {
                    player.sendMessage(new TranslationTextComponent("block.lcc.nether_reactor.y"));
                } else if (!this.checkStructure(world, pos)) {
                    player.sendMessage(new TranslationTextComponent("block.lcc.nether_reactor.incorrect"));
                } else if (!player.isCreative() && !this.checkForPlayers(world, pos)) {
                    player.sendMessage(new TranslationTextComponent("block.lcc.nether_reactor.players"));
                } else {
                    ((NetherReactorTileEntity)world.getTileEntity(pos)).activate();
                    world.setBlockState(pos, state.with(STATE, ReactorState.ACTIVE), 3);
                    player.sendMessage(new TranslationTextComponent("block.lcc.nether_reactor.active"));
                }
            }
            return ActionResultType.SUCCESS;
        }
        return ActionResultType.PASS;
    }

    protected boolean checkStructure(World world, BlockPos pos) {
        BlockPos.Mutable bp = new BlockPos.Mutable();
        for (int i = -1; i <= 1; i++) {
            for (int k = -1; k <= 1; k++) {
                BlockState bottom = world.getBlockState(bp.setPos(pos).move(i, -1, k));
                BlockState middle = world.getBlockState(bp.move(Direction.UP));
                BlockState top = world.getBlockState(bp.move(Direction.UP));
                if (i != 0 && k != 0 && (bottom != BASE || middle != SHELL || !top.isAir(world, pos))) return false;
                if (((i != 0 && k == 0) || (i == 0 && k != 0)) && (bottom != SHELL || !middle.isAir(world, pos) || top != SHELL)) return false;
                if (i == 0 && k == 0 && (bottom != SHELL || top != SHELL)) return false;
            }
        }
        return true;
    }

    protected boolean checkForPlayers(World world, BlockPos pos) {
        AxisAlignedBB range = new AxisAlignedBB(pos.down(), pos.up(2)).grow(9, 0, 9);

        //TODO: config option, big servers won't feasibly get everyone together
        return world.getServer().getPlayerList().getPlayers().stream().allMatch(player -> player.dimension == world.dimension.getType() && range.intersects(player.getBoundingBox()));
    }

    @Override
    public void onBlockPlacedBy(World worldIn, BlockPos pos, BlockState state, LivingEntity placer, ItemStack stack) {
        if (stack.hasDisplayName()) {
            TileEntity tileentity = worldIn.getTileEntity(pos);
            if (tileentity instanceof NetherReactorTileEntity) {
                ((NetherReactorTileEntity)tileentity).setCustomName(stack.getDisplayName());
            }
        }
    }

    public enum ReactorState implements IStringSerializable {
        READY,
        ACTIVE,
        USED;

        @Override
        public String getName() {
            return this.name().toLowerCase();
        }

        @Override
        public String toString() {
            return this.getName();
        }
    }
}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/block/NetherReactorBlock.kt
package com.joshmanisdabomb.lcc.block

import com.joshmanisdabomb.lcc.block.entity.NetherReactorBlockEntity
import com.joshmanisdabomb.lcc.directory.LCCBlockEntities
import com.joshmanisdabomb.lcc.directory.LCCTags
import net.minecraft.block.*
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityTicker
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.state.StateManager
import net.minecraft.state.property.EnumProperty
import net.minecraft.text.TranslatableText
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.StringIdentifiable
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Box
import net.minecraft.util.math.Direction
import net.minecraft.world.World

class NetherReactorBlock(settings: Settings) : BlockWithEntity(settings) {

    init {
        defaultState = stateManager.defaultState.with(reactor_state, NetherReactorState.READY)
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) = builder.add(reactor_state).let {}

    override fun createBlockEntity(pos: BlockPos, state: BlockState) = NetherReactorBlockEntity(pos, state)

    override fun onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hit: BlockHitResult): ActionResult {
        if (state[reactor_state] === NetherReactorState.READY) {
            if (!world.isClient) {
                if (pos.y < world.bottomY + 4 || pos.y >= world.topY - 35) {
                    player.sendMessage(TranslatableText("block.lcc.nether_reactor.y"), true)
                } else if (!this.checkStructure(world, pos)) {
                    player.sendMessage(TranslatableText("block.lcc.nether_reactor.incorrect"), true)
                } else if (!player.isCreative && !this.checkForPlayers(world, pos)) {
                    player.sendMessage(TranslatableText("block.lcc.nether_reactor.players"), true)
                } else {
                    (world.getBlockEntity(pos) as? NetherReactorBlockEntity)?.activate()
                    world.setBlockState(pos, state.with(reactor_state, NetherReactorState.ACTIVE), 3)
                    player.sendMessage(TranslatableText("block.lcc.nether_reactor.active"), true)
                }
            }
            return ActionResult.SUCCESS
        }
        return ActionResult.PASS
    }

    protected fun checkStructure(world: World, pos: BlockPos?): Boolean {
        val bp = BlockPos.Mutable()
        for (i in -1..1) {
            for (k in -1..1) {
                val bottom = world.getBlockState(bp.set(pos).move(i, -1, k))
                val middle = world.getBlockState(bp.move(Direction.UP))
                val top = world.getBlockState(bp.move(Direction.UP))
                if (i != 0 && k != 0 && (!bottom.isIn(LCCTags.nether_reactor_base) || !middle.isIn(LCCTags.nether_reactor_shell) || !top.isAir)) return false
                if ((i != 0 && k == 0 || i == 0 && k != 0) && (!bottom.isIn(LCCTags.nether_reactor_shell) || !middle.isAir || !top.isIn(LCCTags.nether_reactor_shell))) return false
                if (i == 0 && k == 0 && (!bottom.isIn(LCCTags.nether_reactor_shell) || !top.isIn(LCCTags.nether_reactor_shell))) return false
            }
        }
        return true
    }

    protected fun checkForPlayers(world: World, pos: BlockPos): Boolean {
        val range = Box(pos.down(), pos.up(2)).expand(9.0, 0.0, 9.0)

        //TODO: config option, big servers won't feasibly get everyone together
        return world.server!!.playerManager.playerList.stream().allMatch { player -> player.world.dimension === world.dimension && range.intersects(player.boundingBox) }
    }

    override fun getRenderType(state: BlockState) = BlockRenderType.MODEL

    override fun <T : BlockEntity> getTicker(world: World, state: BlockState, type: BlockEntityType<T>): BlockEntityTicker<T>? {
        return if (world.isClient) checkType(type, LCCBlockEntities.nether_reactor, NetherReactorBlockEntity::tick) else checkType(type, LCCBlockEntities.nether_reactor, NetherReactorBlockEntity::tick)
    }

    override fun onPlaced(world: World, pos: BlockPos, state: BlockState, placer: LivingEntity?, itemStack: ItemStack) {
        if (itemStack.hasCustomName()) (world.getBlockEntity(pos) as? NetherReactorBlockEntity)?.customName = itemStack.name
    }

    enum class NetherReactorState : StringIdentifiable {

        READY,
        ACTIVE,
        USED;

        override fun asString() = name.toLowerCase()

    }

    companion object {

        val reactor_state = EnumProperty.of("state", NetherReactorState::class.java)

        fun getMapColor(state: BlockState) = when (state[reactor_state]) {
            NetherReactorState.ACTIVE -> MapColor.BRIGHT_RED;
            NetherReactorState.USED -> MapColor.BLACK;
            else -> MapColor.CYAN;
        }

    }

}